### PR TITLE
Feature/issue 491: Implement search functionality for team members in frontend

### PR DESCRIPTION
**Pull request by KobMiles**
## Github ticket

* [Github ticket #491](https://github.com/orgs/ita-social-projects/projects/64/views/1?filterQuery=assignee%3AKobMiles&pane=issue&itemId=128658747&issue=ita-social-projects%7CVictoryCenter-Back%7C491)

## Description
- Fix any UI inconsistencies and mismatches with figma, related to search
- Implement search functionality on frontend
- Integrate frontend with backend

## How it looks

image
image
image

## Summary of change

- Added a new component that displays the team member as a search item.
- Changed the logic and styles in TeamPage files.
- Wrote/changed tests for the new logic.

## How to recreate changes

- Go to the [Team Page](https://localhost:3000/admin-panel/team)
- Enter your search query
- Click on the item you are interested in from the search list

## CHECK LIST
- [ ]  New tests was added or existing was modified
- [ ]  Changes has severe impact on users
- [ ]  Changes has medium impact on users
- [ ]  Changes has light impact on users
- [ ]  Test covetage was updated
- [ ]  PR meets all conventions

## Summary by CodeRabbit

* **New Features**
  * Name-based team search with live suggestions, 5-item pages, load-more, selection/clear, single-item view, and 2-character minimum.

* **UI**
  * Category-aware toolbar search with rich member items (avatar/initials), item separators, overflow tooltips, mouse-leave focus reset, and refined toolbar layout.

* **API**
  * Paged, cancelable team search integration.

* **Utilities**
  * Added initials generator for display avatars.

* **Tests**
  * Expanded coverage for rendering, tooltip behavior, search flows, paging, selection/clear, aborts, and API parameters.

### CODE REVIEW

File: src/components/admin/search-bar/SearchBar.tsx [hunk 1]
```diff
@@ -258,7 +260,10 @@ export const SearchBar = <T,>({
                         <ul
                             className="search-bar__suggestions-list"
                             ref={suggestionsListRef}
-                            onMouseLeave={hideTooltip}
+                            onMouseLeave={() => {
```
Comment (NovickVitaliy) [inline src/components/admin/search-bar/SearchBar.tsx]: move this to the separate function. Inline functions are considered bad practice

**Review by NovickVitaliy — APPROVED** [review]

File: src/pages/admin/team/components/team-page-toolbar/TeamPageToolbar.scss [hunk 2]
```diff
@@ -1,17 +1,18 @@
...
+    flex: 0 0 38.75rem;
+    min-width: 0;
 }
 
-@media (max-width: 1400px) {
-    .toolbar {
-        flex-wrap: wrap;
-        gap: 0.938rem;
-    }
+.toolbar-actions {
```
Comment (svyat871) [inline src/pages/admin/team/components/team-page-toolbar/TeamPageToolbar.scss]: Suggestion: you can write selectors `&-actions ` and `&-search` inside the selector `.toolbar`

**Review by svyat871 — APPROVED** [review]

File: src/components/admin/search-bar/SearchBar.tsx (hunk 1 above)
Comment (Rominos7) [reply src/components/admin/search-bar/SearchBar.tsx]: Anonymous functions in JSX create a new function on every render. This can have performance implications (especially in large lists or highly dynamic UIs), because it causes React to treat the prop as changed on every render, potentially triggering unnecessary re-renders of child components.

https://legacy.reactjs.org/docs/faq-functions.html#arrow-function-in-render

you can use this

```
const yourFunction = () => {
   callFunction1();
   callFunction2();
}
```


`<YourComponent onClick={yourFunction} />`

File: src/components/admin/search-bar/team-member-search-item/TeamMemberSearchItem.tsx [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+        const nameRef = useRef<HTMLDivElement>(null);
+        const subtitleRef = useRef<HTMLDivElement>(null);
+
+        const categoryName = useMemo(
+            () => categories.find((c) => c.id === item.categoryId)?.name ?? '',
+            [categories, item.categoryId],
+        );
+
+        const imageUrl = useMemo(() => {
+            const img = item.image;
```
Comment (Rominos7) [inline src/components/admin/search-bar/team-member-search-item/TeamMemberSearchItem.tsx]: nit:

`const { image } = item;`

File: src/pages/admin/team/components/team-page-content/TeamPageContent.test.tsx [hunk 4]
```diff
@@ -26,6 +26,19 @@ jest.mock('../team-page-toolbar/TeamPageToolbar', () => ({
             <div data-testid="team-page-toolbar">
                 <button onClick={props.onAddMember}>Add Member</button>
                 <button onClick={() => props.onStatusFilterChange(VisibilityStatus.Published)}>Filter Published</button>
+                <button onClick={props.onSearchLoadMore} data-testid="search-load-more">
+                    Load More Results
+                </button>
+                <button
+                    onClick={() => props.searchItems?.[0] && props.onSearchItemSelect(props.searchItems[0])}
```
Comment (Rominos7) [inline src/pages/admin/team/components/team-page-content/TeamPageContent.test.tsx]: Up to you:
Same stuff with arrow function

Comment (Rominos7) [reply src/pages/admin/team/components/team-page-content/TeamPageContent.test.tsx]: But I assume it is test so maybe leave it as it is

**Review by Rominos7 — COMMENTED** [review]
I have asked Cladue (LLM - Claude sonnet 4.5) to review this changes and what I have received

Up to you to fix them but I thought it would be interesting to post it here

Summary of Changes
The main feature added is a Team Member Search functionality in the admin panel. Here are the key changes:
New Files Added:
[TeamMemberSearchItem.tsx](vscode-webview://00er9m891rqvh4sb4m61bkkg8qselsnijnb1tcftuvdgfg9ki8kc/src/components/admin/search-bar/team-member-search-item/TeamMemberSearchItem.tsx) - New search item component for team members with avatar, name, and category display
[TeamMemberSearchItem.scss](vscode-webview://00er9m891rqvh4sb4m61bkkg8qselsnijnb1tcftuvdgfg9ki8kc/src/components/admin/search-bar/team-member-search-item/TeamMemberSearchItem.scss) - Styling for the search item
[TeamMemberSearchItem.test.tsx](vscode-webview://00er9m891rqvh4sb4m61bkkg8qselsnijnb1tcftuvdgfg9ki8kc/src/components/admin/search-bar/team-member-search-item/TeamMemberSearchItem.test.tsx) - Unit tests for the component
Modified Files:
[SearchBar.tsx](vscode-webview://00er9m891rqvh4sb4m61bkkg8qselsnijnb1tcftuvdgfg9ki8kc/src/components/admin/search-bar/SearchBar.tsx:81) - Added clearActive callback and enhanced mouse leave behavior to clear active index and hide tooltip
[SearchItemWrapper.scss](vscode-webview://00er9m891rqvh4sb4m61bkkg8qselsnijnb1tcftuvdgfg9ki8kc/src/components/admin/search-bar/search-item-wrapper/SearchItemWrapper.scss:17-19) - Added border between adjacent search items
[team.ts](vscode-webview://00er9m891rqvh4sb4m61bkkg8qselsnijnb1tcftuvdgfg9ki8kc/src/const/admin/team.ts:53-56) - Added search configuration constants (page size: 5, min characters: 2)
[main-api.ts](vscode-webview://00er9m891rqvh4sb4m61bkkg8qselsnijnb1tcftuvdgfg9ki8kc/src/const/common/api-routes/main-api.ts:12) - Added new API route for team member search: 'TeamMembers/search'
[TeamPageContent.tsx & TeamPageContent.test.tsx](vscode-webview://00er9m891rqvh4sb4m61bkkg8qselsnijnb1tcftuvdgfg9ki8kc/src/pages/admin/team/components/team-page-content/TeamPageContent.test.tsx) - Extensive updates to implement search functionality with load more, item selection, and comprehensive test coverage
The implementation allows users to search for team members by name with a minimum of 2 characters, displays results with avatars and categories, supports pagination (5 results at a time), and includes proper tooltip handling for overflowing text.

Update Todos

Review TeamMemberSearchItem component implementation

Review SearchBar changes and event handling

Review API integration and constants

Review test coverage and quality

Check for potential bugs and edge cases

Review accessibility and UX considerations
Code Review Summary
I've completed a comprehensive code review of the changes in branch feature/issue-491 against release/1.0.0. Here's my detailed analysis:
✅ Strengths
1. Well-Structured Component Design
The [TeamMemberSearchItem.tsx](vscode-webview://00er9m891rqvh4sb4m61bkkg8qselsnijnb1tcftuvdgfg9ki8kc/src/components/admin/search-bar/team-member-search-item/TeamMemberSearchItem.tsx) component follows React best practices with forwardRef and useImperativeHandle
Good separation of concerns with dedicated SCSS file
Proper use of useMemo for performance optimization
2. Excellent Test Coverage
Comprehensive test suites for both the new component and API layer
Tests cover edge cases (image presence, overflow detection, pagination)
API tests validate parameter handling thoroughly ([team-members-api.test.ts:60-104](vscode-webview://00er9m891rqvh4sb4m61bkkg8qselsnijnb1tcftuvdgfg9ki8kc/src/services/api/admin/team/team-members/team-members-api.test.ts#L60-L104))
3. Good API Design
Clean search API implementation with sensible defaults (offset=0, limit=5)
Proper parameter validation and type safety
Consistent with existing getAll method patterns
4. Accessibility Considerations
Images use empty alt text appropriately (decorative avatars)
aria-hidden="true" on avatar container
Proper keyboard navigation already exists in SearchBar
⚠️ Issues & Concerns
Critical Issues
1. CSS Letter-spacing Invalid Value 🔴
Location: [TeamMemberSearchItem.scss:48, 61, 77, 84](vscode-webview://00er9m891rqvh4sb4m61bkkg8qselsnijnb1tcftuvdgfg9ki8kc/src/components/admin/search-bar/team-member-search-item/TeamMemberSearchItem.scss#L48)
letter-spacing: -2%;  // ❌ Invalid - percentages not supported
Issue: CSS letter-spacing doesn't accept percentage values. This will be ignored by browsers. Fix: Should be:
letter-spacing: -0.02em;  // or -0.32px for 16px font
2. Border Radius Inconsistency 🟡
Location: [TeamMemberSearchItem.scss:14](vscode-webview://00er9m891rqvh4sb4m61bkkg8qselsnijnb1tcftuvdgfg9ki8kc/src/components/admin/search-bar/team-member-search-item/TeamMemberSearchItem.scss#L14)
border-radius: 26.67px;  // Oddly specific value
Issue: Avatar size is 2rem (32px), but border-radius is 26.67px. For a perfect circle, it should be 50% or 1rem. Recommendation: Use border-radius: 50%; for maintainability.
Moderate Issues
3. Initials Edge Case Handling 🟡
Location: [TeamMemberSearchItem.tsx:25-33](vscode-webview://00er9m891rqvh4sb4m61bkkg8qselsnijnb1tcftuvdgfg9ki8kc/src/components/admin/search-bar/team-member-search-item/TeamMemberSearchItem.tsx#L25-L33)
const initials = useMemo(() => {
    return item.fullName
        .split(' ')
        .filter(Boolean)
        .slice(0, 2)
        .map((s) => s[0])  // ⚠️ Could fail if s[0] is undefined
        .join('')
        .toUpperCase();
}, [item.fullName]);
Potential Issues:
Empty names after trim could cause issues
Non-ASCII characters might not uppercase correctly
No fallback for single-character names
Suggested Enhancement:
const initials = useMemo(() => {
    const parts = item.fullName.trim().split(/\s+/).filter(Boolean);
    if (parts.length === 0) return '?';
    return parts
        .slice(0, 2)
        .map((s) => s.charAt(0).toUpperCase())
        .join('');
}, [item.fullName]);
4. Search Query Trimming Inconsistency 🟡
Location: [TeamPageContent.tsx:228-241](vscode-webview://00er9m891rqvh4sb4m61bkkg8qselsnijnb1tcftuvdgfg9ki8kc/src/pages/admin/team/components/team-page-content/TeamPageContent.tsx#L228-L241) The search handler trims the query but the min character check (< 2) happens after. This could lead to confusion:
User types "a " (with space) → trimmed to "a" (length 1) → no search
But visual input shows 2 characters
Recommendation: Clear feedback to users or trim before displaying length requirements.
5. Ref Clearing Pattern 🟡
Location: [SearchBar.tsx:216-220](vscode-webview://00er9m891rqvh4sb4m61bkkg8qselsnijnb1tcftuvdgfg9ki8kc/src/components/admin/search-bar/SearchBar.tsx#L216-L220)
suggestionRefs.current.clear();
searchItems.forEach((item) => {
    const key = getSearchItemKey(item);
    suggestionRefs.current.set(key, React.createRef<SearchItemWrapperRef>());
});
Issue: This runs on every render, creating new refs each time. While not breaking, it's inefficient. Better Pattern: Use a useEffect or only update when items change.
Minor Issues
6. Missing Error Handling in Search 🟢
Location: [TeamPageContent.tsx:259-263](vscode-webview://00er9m891rqvh4sb4m61bkkg8qselsnijnb1tcftuvdgfg9ki8kc/src/pages/admin/team/components/team-page-content/TeamPageContent.tsx#L259-L263) The catch block silently fails:
} catch (e) {
    if (axios.isCancel?.(e) || ...) return;
    setSearchSuggestions([]);  // No user feedback on error
    setHasMoreSearch(false);
}
Recommendation: Consider showing a toast or inline error message for better UX.
7. Magic Number in Border Style 🟢
Location: [SearchItemWrapper.scss:17-19](vscode-webview://00er9m891rqvh4sb4m61bkkg8qselsnijnb1tcftuvdgfg9ki8kc/src/components/admin/search-bar/search-item-wrapper/SearchItemWrapper.scss#L17-L19)
.search-item-wrapper + .search-item-wrapper {
    border-top: 0.5px solid colors.$grey-900;
}
Note: 0.5px borders can render inconsistently across browsers/displays. Consider using 1px with adjusted opacity.
8. Unused statusResetKey in Tests 🟢
Location: [TeamPageContent.test.tsx:46](vscode-webview://00er9m891rqvh4sb4m61bkkg8qselsnijnb1tcftuvdgfg9ki8kc/src/pages/admin/team/components/team-page-content/TeamPageContent.test.tsx#L46)
<span data-testid="status-reset-key">{props.statusResetKey}</span>
The statusResetKey is rendered in the mock but never asserted in tests. Either use it or remove it.
9. Image URL Cache-Busting Side Effect 🟢
Location: [TeamPageContent.tsx:426-427](vscode-webview://00er9m891rqvh4sb4m61bkkg8qselsnijnb1tcftuvdgfg9ki8kc/src/pages/admin/team/components/team-page-content/TeamPageContent.tsx#L426-L427)
if (updatedMember.image && 'url' in updatedMember.image)
    updatedMember.image.url = `${updatedMember.image.url}?cb=${Date.now()}`;
Issue: Mutating the input parameter. While it works, it's a side effect. Better Pattern:
const memberToUpdate = {
    ...updatedMember,
    image: updatedMember.image && 'url' in updatedMember.image 
        ? { ...updatedMember.image, url: `${updatedMember.image.url}?cb=${Date.now()}` }
        : updatedMember.image
};
Code Quality Observations
10. Good Patterns ✅
Consistent use of AbortController for request cancellation
Proper ref management with useRef for values needed in closures
Good memoization strategy with useMemo and useCallback
Comprehensive abort signal handling across all async operations
11. Potential Memory Leak Prevention ✅
The cleanup effect is well-implemented:
useEffect(() => {
    return () => {
        searchAbortControllerRef.current?.abort();
    };
}, []);
📋 Recommendations
High Priority:
Fix CSS letter-spacing - This is a CSS bug that needs immediate correction
Fix border-radius calculation - Use 50% or 1rem for the avatar
Add initials fallback - Handle edge cases for empty/invalid names
Medium Priority:
Improve search error handling with user feedback
Optimize ref creation pattern in SearchBar
Consider trimming feedback in search UX
Low Priority:
Document the image cache-busting strategy
Clean up unused test mock elements
Consider immutable update pattern for edited members
🎯 Overall Assessment
Rating: 7.5/10 This is a solid implementation with good architecture, excellent test coverage, and proper React patterns. The main issues are:
CSS bugs that need fixing (letter-spacing, border-radius)
Minor edge case handling improvements needed
Some UX polish opportunities
The feature appears functionally complete and ready for production after addressing the CSS issues.

File: src/components/admin/search-bar/team-member-search-item/TeamMemberSearchItem.scss [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+
+    &__img {
+        width: 100%;
+        height: 100%;
+        object-fit: cover;
+    }
+
+    &__initials {
+        font-family: fonts.$mulish-regular-font;
+        font-size: 0.875rem;
```
Comment (roman-stozhuk) [inline src/components/admin/search-bar/team-member-search-item/TeamMemberSearchItem.scss]: better use pixels here

File: src/pages/admin/team/components/team-page-content/TeamPageContent.tsx [hunk 6]
```diff
@@ -203,20 +221,100 @@ export const TeamPageContent = () => {
...
+            if (trimmed.length === 0) {
+                searchAbortControllerRef.current?.abort();
+                setSearchSuggestions([]);
+                setHasMoreSearch(false);
+                setIsSearchLoading(false);
+                searchPageRef.current = 0;
+                return;
+            }
+
+            if (trimmed.length < 2) {
```
Comment (roman-stozhuk) [inline src/pages/admin/team/components/team-page-content/TeamPageContent.tsx]: use `TEAM_SEARCH.MIN_CHARACTERS_TO_SEARCH` instead for magic number

**Review by roman-stozhuk — APPROVED** [review]
Good job!

File: src/components/admin/search-bar/team-member-search-item/TeamMemberSearchItem.tsx [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+        const categoryName = useMemo(
+            () => categories.find((c) => c.id === item.categoryId)?.name ?? '',
+            [categories, item.categoryId],
+        );
+
+        const imageUrl = useMemo(() => {
+            return image && 'url' in image ? image.url : null;
+        }, [image]);
+
+        const initials = useMemo(() => {
```
Comment (AndreyBondarchuk) [inline src/components/admin/search-bar/team-member-search-item/TeamMemberSearchItem.tsx]: candidate for util/helper

File: src/components/admin/search-bar/team-member-search-item/TeamMemberSearchItem.tsx [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+            return item.fullName
+                .split(' ')
+                .filter(Boolean)
+                .slice(0, 2)
+                .map((s) => s[0])
+                .join('')
+                .toUpperCase();
+        }, [item.fullName]);
+
+        const getTooltipContent = () => {
```
Comment (AndreyBondarchuk) [inline src/components/admin/search-bar/team-member-search-item/TeamMemberSearchItem.tsx:29]: wanna see at app

File: src/pages/admin/team/components/team-page-content/TeamPageContent.tsx [hunk 9]
```diff
@@ -56,6 +56,10 @@ export const TeamPageContent = () => {
         isEditCategoryModalOpen: false,
         isDeleteCategoryModalOpen: false,
     });
+    const [searchSuggestions, setSearchSuggestions] = useState<TeamMember[]>([]);
+    const [isSearchLoading, setIsSearchLoading] = useState(false);
+    const [hasMoreSearch, setHasMoreSearch] = useState(false);
+    const [selectedSearchMember, setSelectedSearchMember] = useState<TeamMember | null>(null);
```
Comment (AndreyBondarchuk) [inline src/pages/admin/team/components/team-page-content/TeamPageContent.tsx:60]: 14 useState

File: src/pages/admin/team/components/team-page-content/TeamPageContent.tsx
Comment (AndreyBondarchuk) [inline src/pages/admin/team/components/team-page-content/TeamPageContent.tsx:40]: 564 lines file

Comment (AndreyBondarchuk) [inline src/pages/admin/team/components/team-page-content/TeamPageContent.tsx:40]: let's refactor it and after that we can see what we can do with states

**Review by AndreyBondarchuk — CHANGES_REQUESTED** [review]

File: src/components/admin/search-bar/team-member-search-item/TeamMemberSearchItem.tsx (hunk 7 above)
Comment (KobMiles) [reply src/components/admin/search-bar/team-member-search-item/TeamMemberSearchItem.tsx]: Moved initials to a separate util

File: src/components/admin/search-bar/team-member-search-item/TeamMemberSearchItem.tsx (hunk 8 above)
Comment (KobMiles) [reply src/components/admin/search-bar/team-member-search-item/TeamMemberSearchItem.tsx:29]: Not sure what you meant - this code handles overflow tooltips. The getTooltipContent() method is used by parent components via ref to show tooltips when text is too long to fit.

File: src/pages/admin/team/components/team-page-content/TeamPageContent.tsx
Comment (KobMiles) [reply src/pages/admin/team/components/team-page-content/TeamPageContent.tsx:40]: I've extracted the search functionality into a custom hook as requested. The file is now 484 lines. 

For the remaining useState refactoring, I suggest we address that in a separate task after this search feature is completed and merged. This will avoid scope creep and keep the PR focused on search functionality.
